campaignion_loqate/package/src/validator.js

```diff
@@ -0,0 +1,85 @@
+/* global Drupal, jQuery */
+
+import { debounce } from '@moreonion/js-utils/src/utils'
+import { Input } from './input'
+
+var $ = jQuery
+
+/**
+ * Validator for a form containing bank_code and account input elements.
+ *
+ * The elements are assumed to be inside a fieldset.
+ */
+class Validator {
+  constructor (client, settings) {
+    this.client = client
+    this.settings = $.extend({}, settings, {
+      validateTimeout: 300,
+    })
+    this.validating = false
+    this.disabledButtons = $()
+  }
+  bind (wrapper) {
+    this.inputHolder = (new Input($('[name$="[holder]"]', wrapper), Drupal.t('Please enter the account holder’s name.'))).bind()
+    this.inputAccount = (new Input($('[name$="[account]"]', wrapper), Drupal.t('Please enter a valid account number.'))).bind()
+    this.inputSortCode = (new Input($('[name$="[bank_code]"]', wrapper), Drupal.t('Please enter a valid sort code.'), [this.inputAccount])).bind()
+    let validate = debounce(() => {
+      this.validate()
+    }, this.settings.validateTimeout)
+    $(wrapper).on('change keyup', validate)
```

Comment:
This ignores the clientside validation setting of *when* fields should be validated. But I guess this module doesn’t have to be perfect :wink:

Reply:
Where would I find this setting? In general think the module was more “perfect“ before the integration with clientside_validation. I would settle for a good compromise between integration / simplicity / UX in this case.

Reply:
I think the "right" why to do this is registering a custom validator function and let clientside validation call it when it sees fit.

See example on https://git.drupalcode.org/project/clientside_validation/-/blob/7.x-1.47/clientside_validation.api.js#L25 (this expects an equivalent server side validation, without I guess a rule needs to be added manually to jquery validate for the field to use it?)

Reply:
I tried to do that, even by implementing a custom remote validation that works with their API. The problem I got stuck at was on the PHP side: actually making clientside_validate *configure* the validation for these fields based on something in the form. I tried to do this based no form-API #element_validate functions, but I couldn’t make it work as my implementation of [`hook_clientside_validation_rule_alter()`](https://git.drupalcode.org/project/clientside_validation/-/blob/7.x-1.x/clientside_validation.api.php#L71) was never called. webforms seem to be processed entirely differently than normal form-API forms.

Reply:
Well, unless we want to validate server side too, I think this could be done in JS only?

Reply:
Ok, how?

Reply:
Something along the lines of this maybe? (haven’t actually tried it)

```js
$(document).bind('clientsideValidationAddCustomRules', function(event){
    // see http://docs.jquery.com/Plugins/Validation/Validator/addMethod#namemethodmessage
    jQuery.validator.addMethod("myCustomMethod", function(value, element, param) {
        // do the validation
    }, jQuery.format('Error message'));

    // not sure when this needs to happen exactly
    // see https://jqueryvalidation.org/rules/
    $(field).rules("add", {"myCustomMethod": true})
});

```

Reply:
```javascript
    // not sure when this needs to happen exactly
    // see https://jqueryvalidation.org/rules/
    jQuery(field).rules("add", {"myCustomMethod": true})
 ```
 
 Your “not sure …” is exactly what I was asking for. The rest was clear. I think I found a solution by adding the appropriate rule-settings in a `#after_build` callback.

Reply:
Not in the first version of your question…

Maybe there is a "too early" but there doesn’t seem to be a "too late" – when I add the rule in a console on a page it just works.

Reply:
Looks very jQuery validaty now! :laughing:

Reply:
Erm, I was only following your suggestion … :joy: